File: leetcodely/javascript/src/addTwoNumbers.js
/* You are given two non-empty linked lists representing two non-negative integers.
*The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
 *
 * /**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 *
 */

class ListNode{
    constructor(val){
        this.val = val;
        this.next = null;
    }

}
createList = (num) =>{
    let head, curr, prev;
    head = curr = prev = ListNode(-1);
    let strNum = num.toString();
    for(const c of strNum){
        curr = new ListNode(parseInt(c, 10))
        prev.next = curr
    }
    return head;

};

addTwoNumbers = (l1, l2) => {

    let p, curr;
    let carry = 0;
    p = curr = new ListNode(0);
    while (l1 || l2) {
        let a = l1 !== null ? l1.val : 0;
        let b = l2 !== null ? l2.val : 0;
        let sum = a + b + carry;
        curr.next = new ListNode(sum%10);
        curr = curr.next;
        carry = Math.floor(sum/10);
        if(l1 !== null){
            l1 = l1.next;
        }
        if(l2 !== null){
            l2 = l2.next;
        }
    }
    if(carry===1)
        curr.next=new ListNode(carry);
    return p.next;

};
l1 = createList(243);
console.log(l1.val);
l2 = createList(708);
console.log(l2.val);
l3 = addTwoNumbers(l1, l2);
console.log(l3.val);

File: container2/leetcodely/java/src/io/soumasish/AddTwoNumbers.java
package io.soumasish;

public class AddTwoNumbers {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode p = null, head = null;
        int carry = 0;
        while(true){
            int sum = 0;
            if(l1 != null) sum += l1.val;
            if (l2 != null) sum += l2.val;
            sum += carry;
            if(sum >= 10){
                sum = sum - 10;
                carry = 1;
            }else{
                carry = 0;
            }
            ListNode temp = new ListNode(sum);
            if(p == null){
                p = temp;
                head = p;
            }
            else{
                p.next = temp;
                p = p.next;
            }
            if(l1 != null) l1 = l1.next;
            if(l2 != null) l2 = l2.next;
            if(l1 == null && l2 == null && carry == 0) break;

        }
        return head;

    }

}
